design_notes/XDN11 - Error handling.md

```diff
@@ -0,0 +1,194 @@
+
+xlang_propagate_error will be defined in the PAL and will take the following parameters:
+- an IXlangErrorInfo* for the error which is being propagated
+- a xlang_string with the language specific error
+- a xlang_string indicating the error stack if available
+- a xlang_string indicating an identifier for the language
+- an IUnknown pointer with any additional language specific information (can be null)
+
+```
+void xlang_propagate_error(
+	IXlangErrorInfo* errorInfo,
+	xlang_string languageError,
+	xlang_string stack,
+	xlang_string projectionIdentifier,
+	IUnknown* languageInformation)
+```
+
+After calling xlang's origination / propagation API, the projection should return the error as the result so that the consuming projection can receive the error and propagate it to the language it projects.  In some cases, the projection may want to call other functions (i.e. cleanup) before returning the error.  If any of these functions error, the projection may treat them as best effort and ignore the error and still propagate the previous error after calling them.  But, those functions may call the error origination API upon error and override the error info stored on the thread local storage.  To avoid this, the projection should retrieve the error info from the thread local storage before doing the best effort work.  Then after doing that work, it should set the retrieved error info back on the thread local store and then return the error.  To help with this, xlang will provide the xlang_get_error_info and xlang_set_error_info API.
+
+```
+void xlang_get_error_info(IXLangErrorInfo** errorInfo)
+void xlang_set_error_info(IXLangErrorInfo* errorInfo)
+```
+
+## Providing additional language specific information
+
+Most of the parameters for xlang_originate_error and xlang_propagate_error are self-explanatory from the descriptions. But one parameter to call out is the IUnknown parameter used to provide additional language specific information.
+
+A projection can use this parameter to provide additional diagnostic information about the error.  It can do this by passing a COM object implementing the IXlangErrorValue interface.  This interface has a GetValue function which, when implemented, will allow to provide the information as a string.  This design allows any projection that the error propagates through to make this information available without it needing to know the specifics of what additional information each projection provides.
+
+```
+interface IXlangErrorValue : IUnknown
```

Comment:
I'm confused by the relation of this to `IXlangErrorInfo` defined below. Maybe 'Value' is the wrong name for this, and it should be `IXlangErrorExtendedInfo` ?  "Extended info" vs "language info" indicates that it's just generic information and might be related to the runtime of the language vs. the language itself.

Is the intent that a projection could QI for its own `IPythonErrorExtendedInfo` which extends this and has a `GetPyDump` method?

Reply:
As per arch meeting, should this interface inherit from IXlangObject instead of IUnknown

Reply:
This interface was meant as a way to provide custom diagnosic information in a way that allows other projections to expose it without needed to know what is stored.  But, I ended up deciding we don't need to define a interface for this because there is existing xlang interfaces that allow to accomplish this.  But. as noted, a projection can pass something implementing their own interface and query for it when they retrieve this object.